rest_api/limsdb/queries/__init__.py

```diff
@@ -248,42 +249,98 @@ def runs_cst(session, time_since=None, run_ids=None, run_status=None):
-                      t.ArtifactUdfView.udfvalue)\
+                      t.ContainerPlacement.wellxposition, t.ContainerPlacement.wellyposition)\
         .join(t.Process.type)\
         .join(t.Process.protocolstep)\
         .join(t.ProtocolStep.labprotocol)\
         .join(t.Process.processiotrackers)\
-        .join(t.ProcessIOTracker.artifact) \
-        .join(t.Artifact.udfs)\
+        .join(t.ProcessIOTracker.artifact)\
         .join(t.Artifact.containerplacement)\
         .join(t.Artifact.samples)\
         .join(t.Artifact.states)\
         .join(t.Sample.project)\
         .join(t.ContainerPlacement.container)\
-        .filter(t.ProcessType.displayname == 'Eval qPCR Quant')\
-        .filter(t.ArtifactUdfView.udfname.in_(
-            ['Original Conc. (nM)', 'Sample Transfer Volume (uL)', 'TSP1 Transfer Volume (uL)', '%CV',
-             'NTP Volume (uL)', 'Raw CP', 'RSB Transfer Volume (uL)', 'NTP Transfer Volume (uL)', 'Ave. Conc. (nM)',
-             'Adjusted Conc. (nM)', 'Original Conc. (nM)', 'Sample Transfer Volume (uL)',
-             'TSP1 Transfer Volume (uL)']))\
-        .filter(t.ArtifactUdfView.udfvalue != None)
-
-    if library_id:
-        q = q.filter(t.Container.name == library_id)
-    else:
-        if time_from:
-            q = q.filter(t.Process.daterun > func.date(time_from))
-
-        if time_to:
-            q = q.filter(t.Process.daterun < func.date(time_to))
+        .filter(t.ProcessType.displayname == step_name)
+
+    if artifact_udfs:
+        q = q.join(t.Artifact.udfs)\
+            .filter(t.ArtifactUdfView.udfname.in_(artifact_udfs))\
+            .filter(t.ArtifactUdfView.udfvalue != None)
+        q = q.add_columns(t.ArtifactUdfView.udfname, t.ArtifactUdfView.udfvalue)
+
+    if output_udfs:
+        output_artifact = aliased(t.Artifact)
+        q = q.join(t.ProcessIOTracker.output)\
+            .join(output_artifact, t.OutputMapping.outputartifactid == output_artifact.artifactid) \
+            .join(output_artifact.udfs) \
+            .filter(t.ArtifactUdfView.udfname.in_(output_udfs)) \
+            .filter(t.ArtifactUdfView.udfvalue != None)
+        q = q.add_columns(t.ArtifactUdfView.udfname, t.ArtifactUdfView.udfvalue)
+
+    if sample_udfs:
+        q = q.join(t.Sample.udfs)
+        # Allow udfname to be None as some sample will not have any of the UDF
+        if allow_missing_sample_udfs:
+            q = q.filter(or_(t.SampleUdfView.udfname.in_(sample_udfs), t.SampleUdfView.udfname == None))
+        else:
+            q = q.filter(t.SampleUdfView.udfname.in_(sample_udfs)) \
+                .filter(t.SampleUdfView.udfvalue != None)
+        q = q.add_columns(t.SampleUdfView.udfname, t.SampleUdfView.udfvalue)
+
+    if container_name:
+        q = q.filter(t.Container.name == container_name)
+
+    if time_from:
+        q = q.filter(t.Process.daterun > func.date(time_from))
```

Comment:
Should this be >= perhaps, to make it inclusive?

Reply:
I don't think it matters since it's a datetime precise to the second